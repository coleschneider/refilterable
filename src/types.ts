import { History } from 'history';
import { MutableRefObject } from 'react';

export const CREATE_FILTER_MARKER = Symbol("createFilter");

export interface FilterConfig<T = string> {
	parse?(input: string): T; 
  format?(value: T): string; 
  validate?(input: string, parse: (input: string) => T): boolean;
  defaultValue?: T | undefined;
}

export interface FilterObject<T = string> extends FilterConfig<T> {
  paramName: string;
  parse(input: string): T; 
  format(value: T): string; 
  validate(input: string, parse: (input: string) => T): boolean;
}

/**
 * User defined type guard to check if the passed object is a valid instance of FiterObject
 * 
 * @param filter 
 */
export function isFilterObject<T>(filter: any): filter is FilterObject<T> {
  return filter[CREATE_FILTER_MARKER];
}

export type FiltersContextValue = {
  locationObserver: MutableRefObject<{ 
    watch(paramName: string, callback: Function): Function;
    getParamInfo(paramName: string): { hasParam: boolean, paramValue: any };
    getCurrentParams(): URLSearchParams;
  }>;
  history?: History;
  filterRegistry: Map<string, FilterObject<any>>,
}

export type ForwardHistoryAction = 'PUSH' | 'REPLACE';

export type SetFilterOptions = {
  dry?: boolean;
  action?: ForwardHistoryAction;
};

export type FilterSetter<T = string> = (nextValue: T, options?: SetFilterOptions) => string;
