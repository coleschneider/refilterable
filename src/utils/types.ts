import { History } from 'history';
import { CREATE_FILTER_MARKER } from './constants';

export type ParseFunction<T> = (input: string | undefined) => T | undefined;



export interface FilterConfig<T = string> {
	parse?: ParseFunction<T>;
  format?(value: T): string; 
  validate?(input: string, parse: ParseFunction<T>): boolean;
  defaultValue?: string | undefined;
}

export interface FilterObject<T = string> extends FilterConfig<T> {
  paramName: string;
	parse: ParseFunction<T>;
  format(value: T): string; 
  validate(input: string, parse: ParseFunction<T>): boolean;
}

/**
 * User defined type guard to check if the passed object is a valid instance of FiterObject
 * 
 * @param filter 
 */
export function isFilterObject<T>(filter: any): filter is FilterObject<T> {
  return Boolean(filter[CREATE_FILTER_MARKER]);
}

export type FiltersContextValue = {
  locationObserver: { 
    watch(paramName: string, callback: Function): Function;
    getParamInfo(paramName: string): { hasParam: boolean, paramValue: any };
    getCurrentParams(): URLSearchParams;
  };
  history?: History;
  filterRegistry: Map<string, FilterObject<any>>,
}

export type ForwardHistoryAction = 'PUSH' | 'REPLACE';

export type SetFilterOptions = {
  dry?: boolean;
  action?: ForwardHistoryAction;
};

export type FilterSetter<T = string> = (nextValue: T, options?: SetFilterOptions) => string;
export type FilterResetter = (options?: SetFilterOptions) => string;
