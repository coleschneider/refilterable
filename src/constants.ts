import { SetFilterOptions, ForwardHistoryAction } from './types';

export const defaultSetFilterOptions: SetFilterOptions = Object.freeze({
  dry: false,
  action: 'PUSH' as ForwardHistoryAction,
});