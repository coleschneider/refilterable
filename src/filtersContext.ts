import { createContext } from "react";
import { FiltersContextValue } from './types';
import createLocationObserver from "./createLocationObserver";

export default createContext<FiltersContextValue | null>(null);