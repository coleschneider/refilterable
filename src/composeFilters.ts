import invariant from 'invariant';
import { FILTER_COMPOSITION_MARKER } from './utils/constants';
import { FilterObject, FilterComposition, isFilterObject } from './utils/types';

// @todo: support filter compositions in filters
export default function composeFilters(filters: FilterObject<any>[], validate?: (input: any) => boolean): FilterComposition {
  invariant(
    filters && filters.length > 0,
    `re-filter: you called compose filters and didn't pass any filters `
  );

  invariant(
    filters.every(filter => isFilterObject(filter)),
    `re-filter: composeFilters() was called with non-filter objects. 
    Use createFilter() to create filters`,
  );

  return ({
    filters,
    validate: typeof validate === 'function' ? validate : () => true,
    [FILTER_COMPOSITION_MARKER]: true,
  } as FilterComposition) ;
} 