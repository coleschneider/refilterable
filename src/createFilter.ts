import invariant from 'invariant';
import { FilterConfig, FilterObject } from './utils/types';
import { CREATE_FILTER_MARKER } from './utils/constants';


export default function createFilter<T>(paramName: string, config?: FilterConfig<T>): FilterObject<T> {
  invariant(
    String(paramName || '').length > 0,
    `re-filter: param name cannot be empty`,
  );

  const defaultConfig = {
    parse(input: string): any { return input; }, 
    format(value: T): string { return String(value); },
    validate(): boolean { return true; },
    defaultValue: undefined,
    [CREATE_FILTER_MARKER]: true,
  };

  return {
    ...defaultConfig,
    ...config,
    paramName,
  };
}; 