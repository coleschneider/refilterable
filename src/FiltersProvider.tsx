import React, { useEffect, useRef, FunctionComponent } from 'react';
import invariant from 'invariant';
import { History, Location } from 'history';
import filtersContext from './filtersContext';
import { FiltersContextValue, FilterObject } from './types';
import createLocationObserver from './createLocationObserver';
 
type Props = {
  history: History;
  children: JSX.Element[];
}

const FiltersProvider: FunctionComponent<Props> = ({ history, children }: Props) => {
  invariant(
    history,
    `re-filter: FiltersProvider was not passed a history instance. 
     Make sure you pass it via the history prop`,
  );

  const filterRegistry = useRef<Map<string, FilterObject<any>>>(new Map());
  const locationObserver = useRef(createLocationObserver(history.location.search));

  useEffect(() => {
    // listen to location changes
    const unlisten = history.listen((location: Location) => {
      locationObserver.current.notify(location.search);
    });

    // unsubscribe when the component unmounts
    return unlisten;
  }, []);

  const contextValue: FiltersContextValue = { 
    locationObserver,
    history, 
    filterRegistry: filterRegistry.current 
  };

  return (
    <filtersContext.Provider value={contextValue}>
      {children}
    </filtersContext.Provider>
  );
}


export default FiltersProvider;