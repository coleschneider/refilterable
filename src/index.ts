import FiltersProvider from "./FiltersProvider";
import createFilter from "./createFilter";
import useFilter from "./useFilter";
import useReset from "./useReset";

export { useReset, useFilter, createFilter, FiltersProvider };
