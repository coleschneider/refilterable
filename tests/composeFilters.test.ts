import { createFilter, composeFilters } from '../src';
import { isFilterObject, isFilterComposition } from '../src/utils/types';

describe('composeFilters', () => {
  const minFilter = createFilter('min', { 
    parse: parseInt,
    defaultValue: '0',
  });
  
  const maxFilter = createFilter('max', { 
    parse: parseInt,
    defaultValue: '100',
  });

  it('should not work if no filters are passed', () => {
    expect(() => {
      // @ts-ignore
      composeFilters()
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      composeFilters([])
    }).toThrowErrorMatchingSnapshot();
  });

  it('should refuse to work with object literals', () => {
    expect(() => {
      // @ts-ignore
      composeFilters([{}])
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      composeFilters([minFilter, {}])
    }).toThrowErrorMatchingSnapshot();
  });

  it('should return an objet of type FilterComposition', () => {
    const composition = composeFilters([minFilter, maxFilter]);
    expect(isFilterComposition(composition)).toBe(true);
  });

  it('should return accept and pass forward custom validation functions', () => {
    const validate = jest.fn().mockReturnValue(false);
    const composition = composeFilters([minFilter, maxFilter], validate);
    composition.validate({ min: 10, max: 20 });
    expect(validate).toHaveBeenCalledWith({ min: 10, max: 20 });
  });

  test('default validation function should always return true', () => {
    const composition = composeFilters([minFilter, maxFilter]);
    expect(composition.validate({ min: 10, max: 20 })).toBe(true);
    expect(composition.validate({ min: 20, max: 10 })).toBe(true);
  })
});