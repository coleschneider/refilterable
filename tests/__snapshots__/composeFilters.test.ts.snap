// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composeFilters should not work if no filters are passed 1`] = `"re-filter: you called compose filters and didn't pass any filters "`;

exports[`composeFilters should not work if no filters are passed 2`] = `"re-filter: you called compose filters and didn't pass any filters "`;

exports[`composeFilters should refuse to work with object literals 1`] = `
"re-filter: composeFilters() was called with non-filter objects. 
    Use createFilter() to create filters"
`;

exports[`composeFilters should refuse to work with object literals 2`] = `
"re-filter: composeFilters() was called with non-filter objects. 
    Use createFilter() to create filters"
`;
